core/trino-main/src/main/java/io/trino/server/DynamicFilterService.java

```diff
@@ -232,7 +232,13 @@ public boolean isCollectingTaskNeeded(QueryId queryId, PlanFragment plan)
             return false;
         }
 
-        return !getSourceStageInnerLazyDynamicFilters(plan).isEmpty();
+        if (!plan.getPartitioning().equals(SOURCE_DISTRIBUTION)) {
+            // Dynamic filters can be collected by single task only for source distribution (non-fixed) stage.
+            // Such stage must only contain replicated joins.
+            return false;
+        }
+
+        return !getSourceDistributionStageProducedDynamicFilters(plan).isEmpty();
```

Comment:
This changes the logic from `If there's any dynamic filters produced and consumed within a stage` to `If there's any dynamic filters produced by a stage`. Am i correct? Could you please explain why is it necessary? Is it needed to cover a mix of broadcast and partitioned joins?